src/cmdline/mod.rs

```diff
@@ -116,134 +97,98 @@ impl Cmdline {
-        valid_element(v)?;
-        self.has_capacity(k.len() + v.len() + 1)?;
+    pub fn insert_boot_arg<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
+        let s = slug.as_ref().trim();
+        valid_str(s)?;
 
-        self.start_push();
-        self.line.push_str(k);
-        self.line.push('=');
-        self.line.push_str(v);
-        self.end_push();
+        let more_space = s.len()
+            + if !self.boot_args.is_empty() {
+                /* a whitespace should be added first */
+                1
+            } else {
+                0
+            };
+        self.has_capacity(more_space)?;
+
+        if !self.boot_args.is_empty() {
+            // a whitespace should be added first
+            self.boot_args.push(' ');
+        };
+        self.boot_args.push_str(s);
 
         Ok(())
     }
 
-    /// Validates and inserts a key-value1,...,valueN pair into this command line.
+    /// Validates and inserts a string representing an init arg to the current command line.
     ///
     /// # Arguments
     ///
-    /// * `key` - Key to be inserted in the command line string.
-    /// * `vals` - Values corresponding to `key`.
+    /// * `slug` - String representing an init arg that needs to be added to the cmdline.
     ///
     /// # Examples
     ///
     /// ```rust
     /// # use linux_loader::cmdline::*;
     /// # use std::ffi::CString;
     /// let mut cl = Cmdline::new(100);
-    /// cl.insert_multiple("foo", &["bar", "baz"]);
-    /// let cl_cstring = CString::new(cl).unwrap();
-    /// assert_eq!(cl_cstring.to_str().unwrap(), "foo=bar,baz");
-    /// ```
-    pub fn insert_multiple<T: AsRef<str>>(&mut self, key: T, vals: &[T]) -> Result<()> {
-        let k = key.as_ref();
-
-        valid_element(k)?;
-        if vals.is_empty() {
-            return Err(Error::MissingVal(k.to_string()));
-        }
-
-        let kv_str = format!(
-            "{}={}",
-            k,
-            vals.iter()
-                .map(|v| -> Result<&str> {
-                    valid_element(v.as_ref())?;
-                    Ok(v.as_ref())
-                })
-                .collect::<Result<Vec<&str>>>()?
-                .join(",")
-        );
-
-        self.insert_str(kv_str)
-    }
-
-    /// Validates and inserts a string to the end of the current command line.
-    ///
-    /// # Arguments
-    ///
-    /// * `slug` - String to be appended to the command line.
-    ///
-    /// # Examples
-    ///
-    /// ```rust
-    /// # use linux_loader::cmdline::*;
-    /// # use std::ffi::CString;
-    /// let mut cl = Cmdline::new(100);
-    /// cl.insert_str("foobar");
+    /// cl.insert_init_arg("foobar").expect("");
     /// let cl_cstring = CString::new(cl).unwrap();
     /// assert_eq!(cl_cstring.to_str().unwrap(), "foobar");
     /// ```
-    pub fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
-        let s = slug.as_ref();
+    pub fn insert_init_arg<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
```

Comment:
This function looks a lot like the current body of `insert_boot_arg`, which in turn looks somewhat like the old `insert_str`, so I would keep 1 single function for inserting a plain string alongside key-value insertion functions for boot args only.

Reply:
Ok, so if I got it right you suggest to keep 2 functions: one for inserting a string into init_args and the other one for inserting a key-value pair into boot_args. To make it clear for the user, I would suggest to have a naming like `insert_init_args_str()` and `insert_boot_args_pair()`. Do you agree with this naming convention? Otherwise, it would be a little bit confusing for the user, even though the types of the parameters of the functions clarifies the ambiguity.

Reply:
Also, I consider useful to keep a function that inserts a string into boot args as long as we might need to append boot args that are not key-value pairs.

Reply:
Since we already have breaking changes at the interface level, I was thinking about making a more clear separation of concerns between bootargs and initargs by creating 2 separate structures to handle each. For the operations that are the same for both of them, we can re-use the code by implementing the same trait for both structures. For the functionality that is different, we need to define it in the corresponding structure. One example for separate functionality is the `insert_virtio` function which needs to be implemented only for `BootArgs`.

This will result in more code, but I think it makes the interface more logical as well.

```rust
pub struct BootArgs(String);
pub struct InitArgs(String);

pub trait UpdateArgs {
    fn insert_key_value<T: AsRef<str>>(&mut self, key: T, val: T) -> Result<()>;
    fn insert_multiple<T: AsRef<str>>(&mut self, key: T, vals: &[T]) -> Result<()>;
    ......
    /// we need to add all the other common functions here as well.
}

/// To reuse the code we can implement the trait for String and then just have blanket implementations for both InitArgs and BootArgs.
/// There are probably better ways to auto-implement the trait, but we need to think about this a bit more.

#[derive(Clone, Debug)]
pub struct Cmdline {
    boot_args: BootArgs,
    init_args: InitArgs,
}

impl UpdateArgs for BootArgs {...}
impl UpdateArgs for InitArgs {...}

impl BootArgs {
    pub fn insert_virtio_device.....
    /// here we need to add any other functions that are specific to the boot args.
.....
}

impl Cmdline {
     pub fn boot_args_mut(&mut self) -> &mut BootArgs {....}
     pub fn init_args_mut(&mut self) -> &mut InitArgs {...}
}

......
fn how_to_use() {
   let mut cmdline = Cmdline::try_from("some string");
   cmdline.boot_args_mut().insert_virtio_device(.....);
   cmdline.boot_args_mut().insert_multiple(.....);
}



```

Reply:
Working on this